File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq

def solution(jobs):
    answer = 0
    time, idx = 0, 0
    before = -1
    heap = []
    
    while idx < len(jobs):
        for job in jobs:
            if before < job[0] <= time:
                heapq.heappush(heap, [job[1], job[0]])
        if heap:
            cur = heapq.heappop(heap)
            before = time
            time += cur[0]
            answer += (time-cur[1])
            idx += 1
        else:
            time += 1
    return int(answer/len(jobs))

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a, b) -> Integer.compare(a[1], b[1])
        );
        int count = 0;
        int before = -1;
        int time = 0;
        
        while (count < jobs.length) {
            for (int[] job : jobs) {
                if (job[0] <= time && job[0] > before) {
                    pq.add(job);
                }
            }
            
            if (!pq.isEmpty()) {
                int[] cur = pq.poll();
                before = time;
                time += cur[1];
                answer += time-cur[0];
                count++;
            } else {
                time++;
            }
        }
        return answer/jobs.length;
    }
}